### PR TITLE
Pbi 7 1 generate models file

### PR DESCRIPTION
**Pull request by KingOnyxx**
Part of MOT-99
This pull request introduces a significant enhancement to the codebase by adding support for generating Spring Boot-style model templates and enabling bidirectional relationship parsing. Key changes include new methods for Spring Boot template generation, updates to relationship strategies, and extensive test coverage for these features.

### Spring Boot Model Template Generation:
* Added `to_models_springboot_context` in `ClassObject` to generate Spring Boot-specific context for classes, including fields and relationships. (`app/models/diagram.py`, [app/models/diagram.pyR71-R101](diffhunk://#diff-78a5dbdff924e53b14761a27cb0224e46629026c314fecfdec45199126ed0680R71-R101))
* Introduced `to_springboot_models_template` methods in relationship and field classes to map relationships and fields to Spring Boot annotations and types. (`app/models/diagram.py`, [[1]](diffhunk://#diff-78a5dbdff924e53b14761a27cb0224e46629026c314fecfdec45199126ed0680R209-R216) [[2]](diffhunk://#diff-78a5dbdff924e53b14761a27cb0224e46629026c314fecfdec45199126ed0680R242-R252) [[3]](diffhunk://#diff-78a5dbdff924e53b14761a27cb0224e46629026c314fecfdec45199126ed0680R271-R281); `app/models/properties.py`, [[4]](diffhunk://#diff-ac065d402cf0522ea562e6ec590ff895dcfab0d925f65a24f2d2bbb7a4c307b6R129-R136)
* Added `SPRING_TYPE_MAPPING` in `FieldObject` to map field types to Spring Boot-compatible types. (`app/models/properties.py`, [app/models/properties.pyR73-R91](diffhunk://#diff-ac065d402cf0522ea562e6ec590ff895dcfab0d925f65a24f2d2bbb7a4c307b6R73-R91))
* Implemented `print_springboot_style` in `ModelsElements` to render Spring Boot model templates using Jinja. (`app/models/elements.py`, [app/models/elements.pyR109-R136](diffhunk://#diff-8367d83e0338e7cef59037b6e2f43952e60d9fe417368d7e350dfb06fa21f640R109-R136))
* Added a new Jinja template `springboot/model.j2` for Spring Boot model generation. (`app/templates/springboot/model.j2`, [app/templates/springboot/model.j2R1-R40](diffhunk://#diff-6b31ef6d9dc239adcf11b4e80bc16bfb4f5e00d6f68a5738c142974efb6fbba6R1-R40))

### Bidirectional Relationship Parsing:
* Updated `create_relationship` methods in relationship strategies (`OneToOneStrategy`, `ManyToOneStrategy`, `ManyToManyStrategy`) to support bidirectional relationships. (`app/parse_class_pattern/parse_relationship_strategy.py`, [[1]](diffhunk://#diff-b5b1ba6ae42514a43bcdc43ecb1dcfaec50c9e56c7e47ac22e7cc3789545114bL11-R15) [[2]](diffhunk://#diff-b5b1ba6ae42514a43bcdc43ecb1dcfaec50c9e56c7e47ac22e7cc3789545114bL20-R51) [[3]](diffhunk://#diff-b5b1ba6ae42514a43bcdc43ecb1dcfaec50c9e56c7e47ac22e7cc3789545114bR66-R108)
* Modified `parse_relationships` in `ParseJsonToObjectClass` to accept a `bidirectional` flag. (`app/parse_json_to_object_class.py`, [[1]](diffhunk://#diff-c0001b02d7e400a8e68c0efcbb1e4413600f9a1616432b24069f3aece1ecfe05L58-R60) [[2]](diffhunk://#diff-c0001b02d7e400a8e68c0efcbb1e4413600f9a1616432b24069f3aece1ecfe05L78-R82)

### Test Coverage:
* Added unit tests for Spring Boot template generation in `test_model_springboot.py`, covering positive, negative, and edge cases for both classes and fields. (`tests/test_model_springboot.py`, [tests/test_model_springboot.pyR1-R139](diffhunk://#diff-ba334c59511be734240099e7ad13d466f2f7807b19bf2c37c96799f6636cb9ddR1-R139))
* Introduced tests for bidirectional relationship parsing in `test_strategy.py`. (`tests/test_parse_class_pattern/test_strategy.py`, [tests/test_parse_class_pattern/test_strategy.pyR51-R112](diffhunk://#diff-45d4f19e23561106ecda9792f3b63a9fc7d105120d85c136fe0bc245000b6677R51-R112))

These changes enhance the system's flexibility by supporting a new framework (Spring Boot) and improving relationship modeling capabilities.

### CODE REVIEW

File: app/models/diagram.py [hunk 1]
```diff
@@ -68,6 +68,24 @@ def to_models_code_template_context(self) -> dict[str]:
 
         return ctx
 
+    def to_models_springboot_context(self) -> dict[str, dict[str]]:
+        """Returns a dictionary with the class name, parent class name, fields and relationships
+        for the Spring Boot template."""
+        ctx = {
+            "name": self.get_name(),
```
Comment (KenKomKom) [inline app/models/diagram.py:75]: I suggest for name and parent, to use `to_pascal_case` when we are integrating to fit Java's clean code rule.

File: app/models/properties.py [hunk 2]
```diff
@@ -70,6 +70,25 @@ class FieldObject:
         "Decimal": "models.DecimalField(max_digits=10, decimal_places=2)",
     }
 
+    SPRING_TYPE_MAPPING = {
+        "boolean": "boolean",
+        "String": "String",
```
Comment (KenKomKom) [inline app/models/properties.py:75]: I think it's better to move this to utils because the django function that is like this is located there

File: app/models/diagram.py [hunk 3]
```diff
@@ -200,6 +239,17 @@ def to_models_code_template(self) -> dict[str, str]:
         rel_type = f"models.ForeignKey('{name}', on_delete=models.CASCADE)"
         return {"name": f"{name.lower()}FK", "type": rel_type}
 
+    def to_springboot_models_template(self) -> dict[str, str]:
+        source = self.get_source_class().get_name().lower()
+        name = self.get_target_class().get_name()
```
[the remark this answers is not in the document]
Comment (MightyZanark) [reply app/models/diagram.py:237]: I agree with this one, might as well make the variable called target for better readability.

File: app/models/diagram.py [hunk 4]
```diff
@@ -200,6 +232,17 @@ def to_models_code_template(self) -> dict[str, str]:
...
+        if self.get_source_class_own_amount == "1":
```
Comment (MightyZanark) [inline app/models/diagram.py:238]: I think this should be a function call?

File: app/models/diagram.py [hunk 5]
```diff
@@ -218,3 +261,14 @@ def to_models_code_template(self) -> dict[str, str]:
         name = self.get_target_class().get_name()
         rel_type = f"models.ManyToManyField('{name}')"
         return {"name": f"listOf{name.title()}", "type": rel_type}
+
+    def to_springboot_models_template(self) -> dict[str, str]:
+        source = self.get_source_class().get_name().lower()
+        name = self.get_target_class().get_name()
```
Comment (MightyZanark) [inline app/models/diagram.py:267]: Similar comment as the one before.

File: app/templates/springboot/model.j2 [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+import java.util.List;
+import java.util.UUID;
+
+@Entity
+@Getter
+@Setter
+{%- if parent %}
+@Inheritance(strategy = InheritanceType.SINGLE_TABLE)
+{%- else %}
+{% endif -%}
```
Comment (MightyZanark) [inline app/templates/springboot/model.j2:20]: If there is nothing in the else block, might as well remove it.

**Review by MightyZanark — CHANGES_REQUESTED** [review]
Overall looks good, just some nitpicks I would like you to change for better readability.

**Review by DaWanAnOnli — CHANGES_REQUESTED** [review]

File: app/templates/springboot/model.j2 (hunk 6 above)
Comment (KingOnyxx) [reply app/templates/springboot/model.j2]: Thanks for noticing! i have removed the else blocks in jinja

File: app/models/diagram.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (KingOnyxx) [reply app/models/diagram.py]: I have renamed it to target for better readability
